File: 백준/Gold/1744. 수 묶기/수 묶기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Collections;
import java.util.PriorityQueue;


public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int N = Integer.parseInt(br.readLine());
        PriorityQueue<Integer> positive = new PriorityQueue<>(Collections.reverseOrder());
        PriorityQueue<Integer> negative = new PriorityQueue<>();

        int answer = 0;

        for (int i = 0; i < N; i++) {
            int num = Integer.parseInt(br.readLine());
            if (num == 1) {
                answer += 1;
            } else if (num > 1) {
                positive.offer(num);
            } else {
                negative.offer(num);
            }
        }

        while (!positive.isEmpty()) {
            int cur = positive.poll();
            if (!positive.isEmpty()) {
                int next = positive.poll();
                answer += cur * next;
            } else {
                answer += cur;
            }
        }

        while (!negative.isEmpty()) {
            int cur = negative.poll();
            if (!negative.isEmpty()) {
                int next = negative.poll();
                answer += cur * next;
            } else {
                answer += cur;
            }
        }

        System.out.println(answer);
    }

}

File: 백준/Gold/1744. 수 묶기/수 묶기.py
import sys
import heapq
input = sys.stdin.readline

N = int(input())
over = []
under = []

answer = 0

for _ in range(N):
    num = int(input())
    if num == 1:
        answer += 1
    elif num > 1:
        heapq.heappush(over, - num)
    else:
        heapq.heappush(under, num)


while over:
    cur = - heapq.heappop(over)
    if over:
        next = - heapq.heappop(over)
        answer += cur * next
    else:
        answer += cur


while under:
    cur = heapq.heappop(under)
    if under:
        next = heapq.heappop(under)
        answer += cur * next
    else:
        answer += cur


print(answer)
